cpp/src/arrow/dataset/test_util.h

```diff
@@ -555,6 +555,19 @@ class FileFormatFixtureMixin : public ::testing::Test {
     ASSERT_OK_AND_ASSIGN(predicate, predicate.Bind(*full_schema));
     ASSERT_FINISHES_OK_AND_EQ(util::nullopt, fragment->CountRows(predicate, options));
   }
+  void TestFragmentEquals() {
+    auto options = std::make_shared<ScanOptions>();
+    auto test_schema = schema({field("f64", float64())});
+    auto reader = this->GetRecordBatchReader(test_schema);
+    auto other_reader = this->GetRecordBatchReader(test_schema);
+    auto source = this->GetFileSource(reader.get());
+    auto other_source = this->GetFileSource(other_reader.get());
+
+    auto fragment = this->MakeFragment(*source);
+    auto other = this->MakeFragment(*other_source);
+
+    EXPECT_EQ(fragment->Equals(*other), false);
```

Comment:
Also check `fragment->Equals(*fragment)`?

Also use `EXPECT_TRUE` and `EXPECT_FALSE`